### ISSUE TITLE
code quality check: pep8, unused variables, etc.

### ISSUE THREAD

**Issue by dvolgyes** [automl/auto-sklearn#393]
Hi,

Great tool, but it could be improved in several ways.
One of them is to clean up the code a little bit.
I recommend two tools:
- autopep8 for automatic code formatting
- flake8 for static analyzes of the code
(Flake8 can have plugins, I strongly recommend flake8-bugbear which
is automatically detected if installed.)

Install them with pip: 
```
pip install autopep8 flake8 flake8-bugbear
```

There are two worrisome warnings appearing frequently:
- variable assigned but never used
- many python3 incompatibility warnings

And there are some other useful comments, e.g. unused loop variable.

Assigned but never used variables are problematic because
it might be a type, and you might wanted to actually set another
variable. Or it is just an unnecessary line which makes no sense anymore
but makes to code less readable.

If you don't want to use the variable, then using underscore, 
or underscore+variable name is a good pattern., e.g. for _i  in ...

Anyway, I strongly recommend to analyze the code with flake8, 
and fix the non-cosmetic warnings. (Cosmetic: line length, etc., but these
are not important.) (Note that you can ignore warnings in flake8, e.g. 
```
cd auto-sklearn/autosklearn
flake8  ./  --ignore E501,C101,E265 
```

Most fixes are very straightforward, and can be done in a minute, so the code could be clean up in very short time. Clean code also helps to find/fix the non-trivial issues.

**Comment by mfeurer** [comment]
I do 100% agree that the code quality could and should be improved drastically. However, in order to keep the code quality at a consistent level, we need an online service which judges code quality. We've been using landscape.io for quite a while, but it turned out to be unreliable. Do you have any suggestions here? Maybe, we could use a separate task in travis-ci for this? If yes, do you maybe want to contribute a PR which sets this up?

**Comment by dvolgyes** [comment]
Well, it depends.
It is easy to make a Travis job which tests the code with flake8, and terminates with error if it finds anything. However, this is quite annoying. (You have to check Travis log for the errors.) 
On separate sites, there are some tools, like Code Climate, Codacy, etc. 

Unfortunately, I haven't find any which would support flake8. But Codacy seems to contain a large amount of automatization and a good deal of tests with the usual deal: free for open source projects. 

It is highly subjective, but it takes around 2 minutes to set it up, and a few minutes for the first analyzes, so Codacy seems to be the best easy solution for me at this moment. 
I cloned the project, signed it up to Codacy, and this is the default report without specifying any details:
https://www.codacy.com/app/dvolgyes/auto-sklearn

Similarly, HoundCI would comment on pull request.
HoundCI is more like a chatbot which annotates the pull request.
As far as I know HoundCI uses flake8 directly.
https://houndci.com

There is a last one, called Pep8 speaks. 
This might be the most widely used tool, e.g. skimage uses it, and thousands
of other projects. 
It is similar to HoundCI and it comments PEP8 violations.
https://github.com/apps/pep8-speaks/installations/new

Well, it depends. If you feel that you will try to stick to PEP8, I quite recommend all of them.
If you feel overwhelmed, I recommend to skip Codacy, it might be too verbose.

But first, I would run autopep8 formatting, then fixing the remaining flake8 issues by hand, 
and just after that using the above, because at this moment they report every whitespace violation, too long lines, etc., and there are some.

Pull request: codacy could be refined with config file, but it works nicely with the default settings. But you have to sign up because of the permissions for the repository.

**Comment by mfeurer** [comment]
As a first step we're now running flake8, which was added in #562.

**Comment by mfeurer** [comment]
We should also set up [vulture](https://github.com/jendrikseipp/vulture).

**Comment by Borda** [comment]
@mfeurer I have good experience with [CodeFactor](https://www.codefactor.io) and [Codacy](https://www.codacy.com/) where the CodeFactor seems to have faster review response and the comments are slightly easier to understand... Later you can introduce one of those as a required check before merge which would basically deny inserting messy code :)
Moreover, an advantage of using an external app for maintaining quality is that it does not slow down the build/test process on CI.

**Comment by mfeurer** [comment]
Thanks for the suggestion. In our group we had bad experience with relying on external code quality services as so far none worked as reliable as travis-ci. Therefore, we're now building on code from scikit-learn which also does PEP8 checks in travis-ci for the same reason.

**Comment by Borda** [comment]
sure, so you can just have a try and connect one of these as optional 
(I would not suggest to remove pep8 from Travis-CI completely :)
btw, in my personal [repo](https://github.com/Borda/pyImSegm), I do have also both - flake8 in Travis and CodeFactor

**Comment by franchuterivera** [comment]
Hello! Thanks a lot for the suggestion. This has already been implemented in auto-sklearn and we currently enforce these checks via a pre-commit github action.
